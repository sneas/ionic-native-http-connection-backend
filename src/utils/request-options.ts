import { HttpRequest } from '@angular/common/http';
import { HTTP } from '@ionic-native/http/ngx';
import { bodyToObject } from './http-params';
import { collectionToObject } from './collection-to-object';
import { detectSerializer } from './data-serializer';
import { detectResponseType } from './response-type';

type RequestOptions = Parameters<typeof HTTP.prototype.sendRequest>[1];

type HTTPRequestMethod = 'get' | 'post' | 'put' | 'delete' | 'patch' | 'head';

const DATA_REQUEST_METHODS = ['post', 'put', 'patch'];

export const getRequestOptions = (req: HttpRequest<any>): RequestOptions => ({
    method: req.method.toLowerCase() as HTTPRequestMethod,
    headers: collectionToObject(req.headers),
    serializer: detectSerializer(req),
    responseType: detectResponseType(req.responseType),
    ...(!DATA_REQUEST_METHODS.includes(req.method.toLowerCase())
        ? null
        : getRequestData(req)),
});

export const getRequestData = (
    req: HttpRequest<any>,
): Pick<RequestOptions, 'data'> | null => {
    if (
        (req.headers.get('content-type') || '')
            .toLowerCase()
            .indexOf('text/') === 0
    ) {
        return {
            data: req.body,
        };
    }

    return {
        data: bodyToObject(req.body),
    };
};
