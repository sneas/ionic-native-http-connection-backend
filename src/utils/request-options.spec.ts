import { getRequestData, getRequestOptions } from './request-options';
import { HttpHeaders, HttpRequest } from '@angular/common/http';

describe('getRequestOptions', () => {
    it('should skip data param in case of non-data request', () => {
        const request = new HttpRequest<any>('GET', 'http://something.com?a=b');
        expect(getRequestOptions(request)).toMatchSnapshot();
    });

    it('should include data param in case of data request', () => {
        const request = new HttpRequest<any>(
            'POST',
            'http://something.com?a=b',
            {
                a: 'b',
            },
        );
        expect(getRequestOptions(request)).toMatchSnapshot();
    });
});

describe('getRequestData', () => {
    it('returns body data as is in case of text request', () => {
        expect(
            getRequestData(
                new HttpRequest<any>('POST', 'http://something.com', 'a=b', {
                    headers: new HttpHeaders({
                        'content-type': 'text/plain',
                    }),
                }),
            ),
        ).toEqual({
            data: 'a=b',
        });
    });

    it('properly returns data', () => {
        expect(
            getRequestData(
                new HttpRequest<any>('POST', 'http://something.com', 'a=b'),
            ),
        ).toEqual({
            data: {
                a: 'b',
            },
        });
    });
});
