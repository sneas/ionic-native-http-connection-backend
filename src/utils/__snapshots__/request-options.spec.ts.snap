// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRequestOptions should include data param in case of data request 1`] = `
Object {
  "data": Object {
    "a": "b",
  },
  "headers": Object {},
  "method": "post",
  "responseType": "text",
  "serializer": "json",
}
`;

exports[`getRequestOptions should skip data param in case of non-data request 1`] = `
Object {
  "headers": Object {},
  "method": "get",
  "responseType": "text",
  "serializer": "urlencoded",
}
`;
